server/src/main/java/org/opensearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -112,6 +120,13 @@ public String[] concreteIndexNames(ClusterState state, IndicesRequest request) {
         return concreteIndexNames(context, request.indices());
     }
 
+    /**
+     * Returns the registry of system indices that have been reserved by modules and plugins
+     */
+    public SystemIndices systemIndices() {
```

Comment:
I believe the initial design was just exposing system index names? I would be in favor of that versus opening up the API if that's possible. Also it looks like `SystemIndexDescriptor` defines an index pattern. Shouldn't `IndexNameExpressionResolver` be resolving that to an actual index name?

Reply:
The initial design was exposing this [map](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/SystemIndices.java#L81). `Canonical Name of Plugin's main class` -> `Collection<SystemIndexDescriptor>`

Reply:
For the security plugin's use-case it would be better to get the patterns than resolving on every request.

In the SecurityPlugin, it creates a [AccessEvaluator](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/privileges/SecurityIndexAccessEvaluator.java#L83-L85) for requests to system indices and the system index matcher is created in the constructor. There's only one instance of this class per node and its created on bootstrap.

What do you think about a method in this class like?

`public boolean matchesSystemIndex(String... indexPatterns)` where the caller passes index patterns and the system indices are never exposed to the caller?

Reply:
The security plugin does log out the system indices contained in a request if the request is denied.

1. Where system index matching takes place: https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/privileges/SecurityIndexAccessEvaluator.java#L178-L180
2. Audit log and log output containing the name of system indices contained in the request: https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/privileges/SecurityIndexAccessEvaluator.java#L351-L355

Reply:
Could the method signature be `public Set<String> matchesSystemIndex(String... indexPatterns)` and this method returns the matching system indices or empty if it doesn't match any?

Reply:
A signature like `Set<String> concreteSystemIndices(String... indexExpressions)` would match the existing signatures of that class quite well. Do you need to pass in any context like ClusterState that the other methods do?

Also, I do want to note that `SystemIndexPlugin` exposes `SystemIndexDescriptor` yet somehow `SystemIndexDescriptor` does not have the `@PublicApi` annotation. My concerns about increasing the public API scope might not actually be accurate, but I still think the path you're suggesting here might be the better way to go.

Reply:
Security does not require the cluster state. There's class in the Security plugin called [IndexResolverReplacer](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java) that's used to resolve a request and get the list of concrete indices being acted upon.

How its called: https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java#L308

Reply:
@andrross I updated this again to add a method called `public List<String> concreteSystemIndices(String... concreteIndices)` that requires the caller to pass `concreteIndices` instead of `indexExpressions` like other methods in this class. I added a test to show how it can be chained with calls to `concreteIndices` to get the system indices in 2 parts.

I designed it this way since Security has a pretty complex Transport Action -> index resolution process:

The meat of index resolution in the security plugin is:

1. https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java#L215-L352
2. https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java#L616-L808

There's some security specific logic when resolving indices for features like `do_no_fail_on_forbidden` which filters out indices from requests for indices that users are not permitted to perform the specified actions on.

-----

FYI There is a private method in IndexNameExpressionResolver called [checkSystemIndexAccess](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/metadata/IndexNameExpressionResolver.java#L359-L378) which is used [internally](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/metadata/IndexNameExpressionResolver.java#L355) when resolving concrete indices. The only functionality that method provides is logging out that a request matches system indices, but doesn't do anything else. 

FYI Core's logic for determining whether an index is a system index is here: [MetadataCreateIndexService.validateDotIndex](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java#L257-L279)

Reply:
I also ported the [WildcardMatcher](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/support/WildcardMatcher.java) from security to core. If this is added to core, the security plugin can remove this class and refer to the classes from core.